### ISSUE TITLE
sql: limited scan executed using DistSQL results in false positive misplanned ranges

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#101471]
In a test cluster, we found high CPU usage under `DistSQLReceiver.pushMeta`'s call into `RangeCache.Insert`. In aggregate, work related to this was responsible for about 8% of total CPU usage on each node.

<img width="1761" alt="Screenshot 2023-04-13 at 1 16 57 PM" src="https://user-images.githubusercontent.com/5438456/231835607-70d3bd3e-98d9-4e2e-a427-93a79b04ac7c.png">

This was diagnosed to be due to false positive misplaced ranges (`execinfra.MisplannedRanges`) for limited scans. We plan a single table reader for the limited scan ([distsql explain](https://cockroachdb.github.io/distsqlplan/decode.html#eJyUkdFrnEAQxt_7Vwzf0x1suPUO-iAUml4sFYymaklLKsHqIKLn2t1VWo7734taaK6Qa_O28838Zj6-PcJ8b-HC-3wXXPshrW78JE0-BmtKvMDbp1SoobOrcU3v4-iWHn6r41I2I91_8GKPGvo6SLnjNyQpim-8mN59oYYC_9ZPyZEyg0CnSg7zAxu4D3Ag8BqZQK9VwcYoPcnHecgvf8CVAnXXD3aSM4FCaYZ7hK1ty3ARqivVb7YQKNnmdTuPnQTUYP9AxuYVw92dxJPFzuXFaf6t5ZjzkvVGnq1HM75txse-4Z8Q2Kt2OHTGpQYCSZ9Pz428gkA0WJdC1TEEgvpQ2ykAPGfOeYm566rSXOVW6Y1z7m0ffQrTxzi6T1brZ29tz279I-GYTa86w_8VsTxlAlxWvPyiUYMu-E6rYj6zlNHMzULJxi7d3VL43dKaDD6FnYvw9jK8vQjLv-Ds9OpXAAAA___c8fky)). The scan’s maximal span is the entire table, but because of the limit, it only ever touches the first range. DistSQL planning knows this, and so it only creates a single `TableReader`. However, we consider any non-collocated leaseholder for any range in the maximal span to be a misplanned range. There’s a mismatch here between the expectations of physical planning and misplanned range detection.

### Reproduction steps:
```sql
# terminal 1
cockroach demo --insecure --multitenant=false --empty --nodes=9

# terminal 2
cockroach workload init kv --splits=1000

# terminal 1
SET TRACING = ON;
SELECT count(v) FROM [SELECT v FROM kv.kv WHERE k >= -9223353402821271710 ORDER BY k LIMIT 100];
SET TRACING = OFF;

SELECT * FROM [SHOW TRACE FOR SESSION] WHERE location LIKE '%readerbase%';

            timestamp           |       age       |                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                message                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |                                                                                                    tag                                                                                                     |            location            |  operation   | span
--------------------------------+-----------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+--------------+-------
  2023-04-13 16:57:09.422635+00 | 00:00:03.721226 | checking range cache to see if range info updates should be communicated to the gateway                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | [n2,fd4eb130f,distsql.stmt=SELECT count(v) FROM [SELECT v FROM kv.kv WHERE k >= _ ORDER BY k LIMIT _],distsql.gateway=1,distsql.appname=$ cockroach demo,distsql.txn=b0cd8f57-c069-4f01-acd8-f881a75feed0] | sql/execinfra/readerbase.go:53 | colbatchscan |   61
  2023-04-13 16:57:09.423146+00 | 00:00:03.721737 | misplanned ranges: desc: r340:/Table/106/1/-9{204943721096824208-186515405338872608} [(n2,s2):1, (n5,s5):2, (n7,s7):3, next=6, gen=40, sticky=9223372036.854775807,2147483647], lease: repl=(n5,s5):2 seq=5 start=1681405009.542466000,1 epo=1 pro=1681405009.559316000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING, desc: r252:/Table/106/1/-91{86515405338872608-68087089580921008} [(n2,s2):1, (n6,s6):5, (n7,s7):3, next=6, gen=40, sticky=9223372036.854775807,2147483647], lease: repl=(n6,s6):5 seq=5 start=1681405009.086155000,1 epo=1 pro=1681405009.106370000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING, desc: r111:/Table/106/1/-91{49658773822969408-31230458065017808} [(n3,s3):6, (n5,s5):2, (n7,s7):3, next=7, gen=44, sticky=9223372036.854775807,2147483647], lease: repl=(n3,s3):6VOTER_INCOMING seq=5 start=1681405007.854785000,1 epo=1 pro=1681405007.865720000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING, desc: r256:/Table/106/1/-9{112802142307066208-094373826549114608} [(n3,s3):1, (n6,s6):5, (n8,s8):6, next=7, gen=56, sticky=9223372036.854775807,2147483647], lease: repl=(n6,s6):5 seq=7 start=1681405009.926463000,1 epo=1 pro=1681405009.943320000,0, closed_timestamp_policy: LAG_BY_CLUSTER_SETTING, ... | [n2,fd4eb130f,distsql.stmt=SELECT count(v) FROM [SELECT v FROM kv.kv WHERE k >= _ ORDER BY k LIMIT _],distsql.gateway=1,distsql.appname=$ cockroach demo,distsql.txn=b0cd8f57-c069-4f01-acd8-f881a75feed0] | sql/execinfra/readerbase.go:97 | colbatchscan |   61

SELECT count(*) FROM [SHOW TRACE FOR SESSION] WHERE message LIKE '%clearing entries overlapping%';

  count
---------
    928
```

Jira issue: CRDB-26963

**Comment by nvanbenschoten** [comment]
The redundant `RangeCache` updates are also done under write lock, so this is a significant source of mutex contention.

<img width="1761" alt="Screenshot 2023-04-13 at 1 21 08 PM" src="https://user-images.githubusercontent.com/5438456/231836402-8d526951-ed0f-464b-9047-6a9aa5f2b4ff.png">

**Comment by nvanbenschoten** [comment]
When fixing this, we should also update `execinfra.MisplannedRanges` to include the number of misplanned ranges in its truncated trace event. Something like:
```go
diff --git a/pkg/sql/execinfra/readerbase.go b/pkg/sql/execinfra/readerbase.go
index 471094f2204..937be2c761f 100644
--- a/pkg/sql/execinfra/readerbase.go
+++ b/pkg/sql/execinfra/readerbase.go
@@ -94,7 +94,7 @@ func MisplannedRanges(
                        }
                        fmt.Fprintf(&b, "%+v", misplannedRanges[i])
                }
-               log.VEventf(ctx, 2, "misplanned ranges: %s", b.String())
+               log.VEventf(ctx, 2, "%d misplanned ranges: %s", len(misplannedRanges), b.String())
        }

        return misplannedRanges
```

**Comment by yuzefovich** [comment]
I'm going to mark this as GA-blocker so that we can get the fix into the next beta in order to see that this CPU usage goes away on 23.1 test cluster.

**Comment by yuzefovich** [comment]
Re-opening till 23.1.0 backport is merged.

**Comment by yuzefovich** [comment]
Fixed by #101702.
